api/src/main/java/org/jenkinsci/plugins/workflow/graph/FlowGraphWalker.java

```diff
@@ -53,4 +56,59 @@ public FlowNode next() {
         }
         return null;
     }
+
+    @Override
+    public Iterator<FlowNode> iterator() {
+        return new FlowGraphWalkerIterator();
+    }
+
+    private class FlowGraphWalkerIterator implements Iterator<FlowNode> {
+
+        private FlowNode next;
+        private int expectedCount;
+
+        public FlowGraphWalkerIterator() {
+            expectedCount = q.size();
+            next = getNext();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return next != null;
+        }
+
+        @Override
+        public FlowNode next() {
+            FlowNode temp = next;
+            next = getNext();
+            return temp;
+        }
+
+        @Override
+        public void remove() {
+            getNext();
+        }
+
+        private FlowNode getNext() {
+            checkForComodification();
+            while (!q.isEmpty()) {
+                FlowNode n = q.pop();
+                if (!visited.add(n)) {
+                    continue;
+                }
+
+                List<FlowNode> parents = n.getParents();
+                addHeads(parents);
+                expectedCount = q.size();
+                return n;
+            }
+            return null;
+        }
+
+        private final void checkForComodification() {
+            if (q.size() != expectedCount) {
+                throw new ConcurrentModificationException();
```

Comment:
@tfennelly it ought to be tested, but I believe this is not an issue. The flow graph is append-only: you can create new heads, but not modify the structure of existing nodes. Since `FlowGraphWalker` takes a snapshot of the current heads in its constructor, if new steps run while you are iterating they simply will not be seen. Akin to a `CopyOnWriteWhatever` collection.